File: The_Algorithm_Design_Manual/4_Sorting_and_Searching/binary_search.c
#include <stdio.h>

int binarySearch(int* arr, int key, int low, int high) {
    if (low > high) return -1;

    int middle = (low + high) / 2;

    if (arr[middle] == key) return middle;

    if (arr[middle] < key) {
        return binarySearch(arr, key, middle+1, high);
    } else {
        return binarySearch(arr, key, low, middle-1);
    }
}

int main(int argc, char* argv[]) {
    #define SIZE 5
    int arr[SIZE] = {1,2,3,4,5};

    int i;
    for(i=0;i<SIZE;i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    printf("Search %d: %d\n", 5, binarySearch(arr, 5, 0, SIZE-1));
    printf("Search %d: %d\n", 9, binarySearch(arr, 9, 0, SIZE-1));

    return 0;
}

File: The_Algorithm_Design_Manual/4_Sorting_and_Searching/BinarySearch.java
public class BinarySearch {
    public static int binarySearch(int[] arr, int key) {
        if (arr == null) return -1;

        int low = 0;
        int high = arr.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] == key) {
                return mid;
            } else if (arr[mid] < key) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        int[] arr = new int[]{1, 2, 3, 4, 5};
        
        System.out.print("Array: ");
        for (int v : arr) {
            System.out.print(v + ", ");
        }
        System.out.println();

        System.out.printf("Search %d: %d\n", 5, binarySearch(arr, 5));
        System.out.printf("Search %d: %d\n", 9, binarySearch(arr, 9));
    }
}
